lang/julia/Portfile

```diff
@@ -22,124 +22,82 @@ long_description    Julia is a high-level, high-performance dynamic programming
 
 github.tarball_from releases
 
-distfiles           ${name}-${version}-full${extract.suffix}
+distfiles           ${name}-${version}-full${extract.suffix} \
+                    ${name}-${version}-full${extract.suffix}.asc
 
-checksums           rmd160  c5a50b478cbef04ce2937c46173ebc63d9403934 \
-                    sha256  1e34c13091c9ddb47cf87a51566d94a06613f3db3c483b8f63b276e416dd621b
+checksums           ${name}-${version}-full${extract.suffix} \
+                    rmd160  8eac1e29a71bac430c93810afac15be07c8ed62b \
+                    sha256  3c5395dd3419ebb82d57bcc49dc729df3b225b9094e74376f8c649ee35ed79c2 \
+                    size    96079453 \
+                    ${name}-${version}-full${extract.suffix}.asc \
+                    rmd160  5b43a22ac7c852639fb251cc94a16c012efd7629 \
+                    sha256  4553930eae7745123dc6229557c8ace54b1a1e4f4f1d979e87d3578593581546 \
+                    size    866
 
-# Force buildbot to skip this since julia builds its own blas; see https://trac.macports.org/ticket/45187
+# Force buildbot to skip this since julia builds its own blas; see #45101
 archive_sites
 
-set llvm_ver        4.0
+depends_fetch-append \
+                    port:gnupg2
 
-depends_lib-append  port:llvm-${llvm_ver} \
+depends_lib-append \
+    port:arpack \
     port:curl \
-    port:pcre2 \
+    port:fftw-3 \
+    port:fftw-3-single \
     port:gmp \
+    port:libgit2 \
     port:mpfr \
+    port:ncurses \
+    port:pcre2 \
     port:SuiteSparse_UMFPACK \
     port:SuiteSparse_CHOLMOD \
     port:SuiteSparse_AMD \
     port:SuiteSparse_CAMD \
     port:SuiteSparse_COLAMD \
-    port:SuiteSparse_SPQR \
-    port:fftw-3 \
-    port:fftw-3-single \
-    port:ncurses \
-    port:arpack \
-    port:libgit2 \
-    path:lib/libopenblas.dylib:OpenBLAS
+    port:SuiteSparse_SPQR
+
+post-checksum {
+    # check GPG signature: https://julialang.org/juliareleases.asc
+    # pre-load public key to avoid keyserver downtime issues
+    # note: tcl exec will return error if error messages not directed to /dev/null
+    system -W ${workpath} "/bin/sh -c '${prefix}/bin/gpg --homedir ${workpath}/.gnupg --import ${filespath}/juliareleases.asc 2>/dev/null || /usr/bin/true'"
+    set gpg_verification [exec /bin/sh -c "if ${prefix}/bin/gpg --homedir ${workpath}/.gnupg --verify ${distpath}/${name}-${version}-full${extract.suffix}.asc ${distpath}/${name}-${version}-full${extract.suffix} 2>/dev/null; then echo 'VERIFIED'; else echo 'UNVERIFIED'; fi"]
+    if {[string trim ${gpg_verification}] != "VERIFIED"} {
+        ui_error "GPG signature verification failed with key juliareleases.asc."
+    }
+}
```

Comment:
There is no reason to check the GPG signature if you have done it once with a file that you used to generate the checksums. No one will be able to forge the checksums, so no one can alter those files and still have Macports download them.

Reply:
I disagree.

Maybe I made a mistake. Maybe I forgot. Maybe some future maintainer made a mistake or forgot. Checking the signature at install time addresses these possibilities.  Why have strong signature verification at all if it's not used, and publicly shown to be used?

Reply:
What is unnecessary is checking hashes against the `.asc` signature file.

Is there a MacPorts way of skipping hash checks on the `.asc` file, as it is only used for gpg signature verification?

Reply:
It's good to check signatures, but unnecessary to check them if someone else has checked them and there's validation that the file is known good. As for adding it to portfiles, gpg might not even be installed, and right now the PGP key servers are in the middle of disintegrating because of a denial of service attack. I'm not sure this is the time to be adding dependencies on it.

Reply:
The concept of gpg signature verification you describe is highly insecure.

It is necessary for installers or MacPorts buildbots to verify gpg signatures independently, not trust that me or you or anyone else to have verified the signature.

The existence of DDOS is irrelevant. Whether DDOS is occurring or not, the author’s public key is now part of the submission, so checking signatures cannot be affected by the availability of key servers.

Reply:
I tried removing the unnecessary `checksums` for the `.asc` file, but `port` requires these for all fetched downloads.

The bottom line is that `port` requires the checksums, and independent author signature verification is required at install time, so the approach used in this Portfile is what's necessary to use secure gpg signatures.

Reply:
> The concept of gpg signature verification you describe is highly insecure

You'll have to do better than asserting.

> It is necessary for installers or MacPorts buildbots to verify gpg signatures independently, not trust that me or you or anyone else to have verified the signature.

Consider that the user has to trust that the Portfile author isn't trying to take over their machine already. There are pretty horrible threat models if you can't trust the Portfile author.

Anyway, let me be blunt: I do security work for a living, I wrote the PGP 2.0 documentation, I'm one of the originators of the (badly designed, I'm not proud of it) PGP web-of-trust model, and I think I understand the threat model just a touch here. None of this is strong evidence that I'm right, but it is a claim that I *have* thought about this set of problems before. You'll have to do a lot better persuading me that a single portfile should have a hack to check a PGP signature when there's no possibility that the file could be different from the one the original Portfile author used.

Reply:
Anyway: *if* we're going to check digital signatures on distribution files, the code for that should be centralized and not in individual portfiles. I'm also far from convinced (to say the least) that it's good or necessary, but it certainly shouldn't be done ad hoc portfile by portfile.

Reply:
Yes, it should be.

But it’s not now, so this is out-of-scope for bringing this important package up to date.

If there’s ever a centralized approach, it can easily be incorporated within the Portfiles.

Reply:
@pmetzger 
> There are pretty horrible threat models if you can't trust the Portfile author. … check a PGP signature when there's no possibility that the file could be different from the one the original Portfile author used.

You’re conflating trust in the validity of the downloads with trust that the downloads were signed by the source author.

Because it’s trivial to verify the signature at install time, there’s no reason to trust the Portfile author to have performed this check correctly. It’s quite easy to mess up or ignore entirely, given the complex nature of PGP, as you mention. 

The package authors have provided a PGP signature to verify their package. Are you arguing that we shouldn’t verify this author-provided signature?

If you’re arguing to check author-provided signatures, just run `gpg --verify` during the install, and don’t leave everyone wondering if the Portfile authors did this correctly, which we inevitably will not.

Reply:
I think part of the complaint here is the logic to perform a PGP check should not be part of the portfile itself. Note though, this does not mean it should be be macports base either. It could probably be something put into a port group, which then any port like julia that wishes to do this could use. We should not commit something that, by design, will force ports to duplicate the same code once more than one port needs it.

Port groups live in the same github project as the ports, so can be modified / added just as easily. see

https://github.com/macports/macports-ports/tree/master/_resources/port1.0/group

Reply:
This issue is out-of-scope.

The Macports version of Julia was released in 2017. It’s 2019. The Julia team is explicitly telling people not to use package mangers like MacPorts because they’re out-of-date and broken, like this port is.

May we please focus limited resources on updating Julia, not where a `gpg --verify` command lives?

Personally, I have neither the knowledge to write a port group to check the Julia packages PGP signature, now the time resources.

I’d be happy to modify this Portfile during the next Julia version update to incorporate any such capability if it exists then.

Reply:
Per @pmetzger and @neverpanic's comments, I've removed PGP signature verification.

However, I kept the code in an `if` block to use the next time I have to check the signature. I'm pretty sure I'll forget how to do it correctly, lose the necessary file locations, or something else, so I believe it's worth keeping this information around.

I still believe that it's worth doing automatically in the Portfile or equivalent, but am happy to defer to consensus opinion.